Side-effect imports are checked if file uses JSX: unintentional breaking change?

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

Commit 7a71887c23a110009bc974f626245f03066e6926 (#27340) appears to have introduced an unintentional breaking change: side-effect module imports are now checked, but only if the file uses JSX.  This is currently causing the DefinitelyTyped CI (which uses `typescript@next`) to fail on `@types/react-table`.

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  7a71887c23a110009bc974f626245f03066e6926 (problem does not occur in 12e371beaaa67a65a79e18a3a20244c4cd102044)

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** side effect module import jsx

**Code**
```ts
// code.tsx
import * as React from "react";
import "nope";
<div/>;
// tsconfig.json
{
    "compilerOptions": {
        "jsx": "react",
        "strict": true
    }
}
// package.json
{
  "devDependencies": {
    "@types/react": "^16.7.6"
  }
}
```

**Expected behavior:** No compile error.

**Actual behavior:** `error TS2307: Cannot find module 'nope'.`  But if the JSX element is removed from the file, the error goes away.

**Playground Link:** N/A, React dependency

**Related Issues:** #28032, etc. for the previous behavior

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -2238,8 +2238,8 @@ namespace ts {
             return initializer || decl;
         }
 
-        function resolveExternalModuleName(location: Node, moduleReferenceExpression: Expression): Symbol | undefined {
-            return resolveExternalModuleNameWorker(location, moduleReferenceExpression, Diagnostics.Cannot_find_module_0);
+        function resolveExternalModuleName(location: Node, moduleReferenceExpression: Expression, ignoreErrors?: boolean): Symbol | undefined {
+            return resolveExternalModuleNameWorker(location, moduleReferenceExpression, ignoreErrors ? undefined : Diagnostics.Cannot_find_module_0);
         }
 
         function resolveExternalModuleNameWorker(location: Node, moduleReferenceExpression: Expression, moduleNotFoundError: DiagnosticMessage | undefined, isForAugmentation = false): Symbol | undefined {
@@ -2601,7 +2601,7 @@ namespace ts {
                 // Try to make an import using an import already in the enclosing file, if possible
                 for (const importRef of containingFile.imports) {
                     if (nodeIsSynthesized(importRef)) continue; // Synthetic names can't be resolved by `resolveExternalModuleName` - they'll cause a debug assert if they error
-                    const resolvedModule = resolveExternalModuleName(enclosingDeclaration, importRef);
+                    const resolvedModule = resolveExternalModuleName(enclosingDeclaration, importRef, /*ignoreErrors*/ true);
                     if (!resolvedModule) continue;
                     const ref = getAliasForSymbolInContainer(resolvedModule, symbol);
                     if (!ref) continue;

--- a/tests/baselines/reference/jsxImportForSideEffectsNonExtantNoError.js
+++ b/tests/baselines/reference/jsxImportForSideEffectsNonExtantNoError.js
@@ -1,0 +1,16 @@
+//// [jsxImportForSideEffectsNonExtantNoError.tsx]
+/// <reference path="/.lib/react16.d.ts" />
+import * as React from "react";
+
+import "./App.css"; // doesn't actually exist
+
+const tag = <div></div>;
+
+
+//// [jsxImportForSideEffectsNonExtantNoError.js]
+"use strict";
+exports.__esModule = true;
+/// <reference path="react16.d.ts" />
+var React = require("react");
+require("./App.css"); // doesn't actually exist
+var tag = React.createElement("div", null);

--- a/tests/baselines/reference/jsxImportForSideEffectsNonExtantNoError.symbols
+++ b/tests/baselines/reference/jsxImportForSideEffectsNonExtantNoError.symbols
@@ -1,0 +1,12 @@
+=== tests/cases/compiler/jsxImportForSideEffectsNonExtantNoError.tsx ===
+/// <reference path="react16.d.ts" />
+import * as React from "react";
+>React : Symbol(React, Decl(jsxImportForSideEffectsNonExtantNoError.tsx, 1, 6))
+
+import "./App.css"; // doesn't actually exist
+
+const tag = <div></div>;
+>tag : Symbol(tag, Decl(jsxImportForSideEffectsNonExtantNoError.tsx, 5, 5))
+>div : Symbol(JSX.IntrinsicElements.div, Decl(react16.d.ts, 2420, 114))
+>div : Symbol(JSX.IntrinsicElements.div, Decl(react16.d.ts, 2420, 114))
+

--- a/tests/baselines/reference/jsxImportForSideEffectsNonExtantNoError.types
+++ b/tests/baselines/reference/jsxImportForSideEffectsNonExtantNoError.types
@@ -1,0 +1,13 @@
+=== tests/cases/compiler/jsxImportForSideEffectsNonExtantNoError.tsx ===
+/// <reference path="react16.d.ts" />
+import * as React from "react";
+>React : typeof React
+
+import "./App.css"; // doesn't actually exist
+
+const tag = <div></div>;
+>tag : JSX.Element
+><div></div> : JSX.Element
+>div : any
+>div : any
+

--- a/tests/cases/compiler/jsxImportForSideEffectsNonExtantNoError.tsx
+++ b/tests/cases/compiler/jsxImportForSideEffectsNonExtantNoError.tsx
@@ -1,0 +1,7 @@
+// @jsx: react
+/// <reference path="/.lib/react16.d.ts" />
+import * as React from "react";
+
+import "./App.css"; // doesn't actually exist
+
+const tag = <div></div>;

